### ISSUE TITLE
twisted.internet.process.PTYProcess throws "OSError: [Errno 6] Device not configured" on Mountain Lion

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5880]
|<img alt="faulkner's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| faulkner reported|
|-|-|
|Trac ID|trac#5880|
|Type|defect|
|Created|2012-08-19 18:19:41Z|
|Branch|https://github.com/twisted/twisted/tree/TIOCNOTTY-5880|

Easy to verify this by running trial.
```
example:
ManholeLoopbackStdio
    testClassDefinition ... Upon execvpe /usr/local/bin/python ['/usr/local/bin/python', '/Users/chris/svn/twisted/twisted/conch/stdio.py', 'twisted.conch.stdio.ConsoleManhole'] in environment id 140257669941472
:Traceback (most recent call last):
  File "/Users/chris/svn/twisted/twisted/internet/process.py", line 418, in _fork
    self._setupChild(**kwargs)
  File "/Users/chris/svn/twisted/twisted/internet/process.py", line 1001, in _setupChild
    os.close(fd)
OSError: [Errno 6] Device not configured
```

<details><summary>Searchable metadata</summary>

```
trac-id__5880 5880
type__defect defect
reporter__faulkner faulkner
priority__normal normal
milestone__ 
branch__branches_TIOCNOTTY_5880 branches/TIOCNOTTY-5880
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__osx osx
time__1345400381000000 1345400381000000
changetime__1347699707000000 1347699707000000
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Worth noting that, as a result of this error, the test suite hangs.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

FreeBSD and OS X both have this in 'man 4 pty':

```
TIOCNOTTY void
    This call is obsolete but left for compatibility.
```

Linux indicates that the call is still used, and perhaps it's still used on other platforms too, but fixing it causes no tests to fail and allows the existing suite to continue.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Automation didn't associate a branch; not sure why.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Since I did this at a sprint I forced a build on `twisted.internet.test.test_process` and `twisted.test.test_process` separately; they're here: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/TIOCNOTTY-5880&num_builds=2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

> Automation didn't associate a branch; not sure why.

1.4.1 (in the second list) on http://twistedmatrix.com/trac/wiki/CommitterCheckList covers this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @glyph|
|-|-|

glyph elaborated on irc:

    We want to disassociate from the controlling terminal in the subprocess as part of `spawnProcess(usePTY=True)`.  In Linux, in order to do that, my understanding is that you need to use both `TIOCNOTTY` and `TIOCSCTTY` in order to do that.  You have "never" (for Twisted-relevant durations of "never") needed to specify `TIOCNOTTY` on BSD-like platforms to do that, but it generally doesn't do any harm.  Now, in Mountain Lion, there's a bug that causes `close()` to blow up after `TIOCNOTTY` is used.  This causes PTY support to break.  The bug isn't ours, but we can work around it trivially by avoiding the call to `TIOCNOTTY` which was previously a no-op on this platform  anyway.  There are other ways to fix it (like ignoring the error on `close()`) but since just skipping several pointless syscalls doesn't cause the tests to break, I thought that would be a better way to go.

Which seems sensible to me.  However, I pointed out:

    What I was trying to say is that there are no unit tests for any of the `TIOCNOTTY` related code.  There are unit tests that accidentally invoke that code, but apart from depending on that code not raising random exceptions, they don't depend on the behavior provided in any way.  So, my idea is... allocate a pty, launch a process with it.  Have that process launch a process with `spawnProcess(usePTY=True)`.  assert that the child and the grand child have different ttys.

This sounds like a good idea to me, but of course I haven't actually tried this approach myself, and terminals are often surprising.  Hopefully this or at least a similar testing approach will give us good coverage of the desired behavior, though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

[ticket:2779 Apparently this chunk of code has also caused problems on AIX in the past, although in a slightly different way.]

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Further research suggests that, in fact, this code is pointless on all platforms.

We _do_ have test coverage of the intended effect - that the controlling terminal, also known as `/dev/tty`, is attached to the file descriptor synthesized by `spawnProcess`, the one that your `ProcessProtocol.transport.write` will write to.  And, if all the code that makes this happen is deleted, `twisted.test.test_process.PosixProcessTestCasePTY.testOpeningTTY` even _fails_!  (On Linux I do have to hit 'enter' to get it to fail rather than hang, so the test could surely be better, but it _is_ present.)

But I can delete all the TIOCNOTTY code and it makes no difference anywhere.

[Quoth the POSIX](http://pubs.opengroup.org/onlinepubs/009695399/functions/setsid.html):

    "Upon return the calling process ... shall have no controlling terminal."

In order to achieve the desired, and already tested-for effect, all that is necessary is to (A) call `setsid()`, and then (B) call `ioctl(slavefd, TIOCSCTTY)`, which we already do after all the shenanigans with `TIOCNOTTY`.  Effectively, what this code is doing right now is starting to daemonize itself, then changing its mind and giving itself a different controlling terminal instead.

So, I'm just going to delete the TIOCNOTTY code instead of what the branch currently does.  And, I posit that no new tests are required because it's totally a no-op!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

(In [35647]) It turns out that none of this is ever necessary, nor has it ever been necessary.  Refs [#5880](https://github.com/twisted/twisted/issues/5880)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph removed owner|
|-|-|

[I forced a build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/TIOCNOTTY-5880), and it looks good aside from unrelated issues on the XP 32 builder and a red herring on the documentation builder.  I think this can land:
```
$ diffstat
 process.py |   17 ++---------------
 1 file changed, 2 insertions(+), 15 deletions(-)
```

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince set owner to @tomprince|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince set owner to @glyph|
|-|-|

* I can confirm that deleting the TIOCSCTTY code (and not the TIOCNOTTY code) cause a test failure in twisted.test.test_process.PosixProcessTestCasePTY.testOpeningTTY (on linux), so the change is adequately covered by tests.
* That test checks that reading/writing to the controlling terminal in the subprocess gets relayed back to the parent. I can't see anyway for this to happen without trial's tty being hooked up to itself, and even then it is likely to break or hang.
* I assume this fix has been tested on Mountain Lion.
* `_setupChild` could perhaps use an expand docstring.
* There is no topfile.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Thanks for the review, Tom.  This review is in a bit of a weird format though. You might want to re-read the ["how to be a good reviewer"](http://twistedmatrix.com/trac/wiki/ReviewProcess#Howtobeagoodreviewer) section of the review process documentation.  In particular I'm not sure if you want me to commit this or submit it for re-review.  You also haven't really separated out your subjective impressions from the required responses to review feedback, which is always helpful.

Replying to [tom.prince](#note_13):
> * I can confirm that deleting the TIOCSCTTY code (and not the TIOCNOTTY code) cause a test failure in twisted.test.test_process.PosixProcessTestCasePTY.testOpeningTTY (on linux), so the change is adequately covered by tests.

Thanks for verifying that.  I guess I don't have to do anything there.

> * That test checks that reading/writing to the controlling terminal in the subprocess gets relayed back to the parent. I can't see anyway for this to happen without trial's tty being hooked up to itself, and even then it is likely to break or hang.

... yes ...

> * I assume this fix has been tested on Mountain Lion.

Yes, it has; I should have stated that explicitly.

> * `_setupChild` could perhaps use an expand docstring.

OK.

> * There is no topfile.

Uh... no? [Here's the topfile in the source browser.](http://twistedmatrix.com/trac/browser/branches/TIOCNOTTY-5880/twisted/topfiles/5880.bugfix)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

Better docstring here: http://twistedmatrix.com/trac/changeset/35693/branches/TIOCNOTTY-5880

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Talked to therve about this and he suggests it should be construed as a passing review.  So I'll treat it that way.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set status to `closed`|
|-|-|

(In [35696]) Merge TIOCNOTTY-5880.

Author: glyph

Reviewer: exarkun, tom.prince

Fixes: [#5880](https://github.com/twisted/twisted/issues/5880)

Remove useless code that redundantly invokes `ioctl(TIOCNOTTY)` before calling `setsid`, that was causing problems on OS X 10.8.
